### ISSUE TITLE
New Feature: Do not backup files present in a backuped archive

### ISSUE THREAD

**Issue by besoft** [duplicati/duplicati#2317]
I have:
- [x] searched open and closed issues for duplicates

----------------------------------------

### Version info

**Duplicati Version:** 2.1.0.0
**Operating System:**
**Backend:**

### Problem
It happens quite often that we have an archive (.ZIP, .7z, .RAR, .tar.gz, etc.) in one directory (say D:\Programs\Packages) and one or more files from this archive extracted in another directory (say D:\Programs\Sources). For example this happens if you develop an open source software using external libraries whose source code is present in some package (and you may have multiple versions to support multiple platforms), or if you have some large collection of data (stored in an archive) from which you are currently using only a subset (extracted on your disk).

When you backup the root directory (e.g., D:\Programs), Duplicati transfers both the archive and unpacked files, which is a waste of space. 

### Feature suggestion
1. When backup starts, the files with extensions for supported archives (.7z, .zip, .tar, .gz, ...) are processed first, all the other files are processed afterwards
2. An archive is processed as a normal file, and then its files are extracted and hash blocks are created for each of this file, perhaps with some special prefix identifying the archive and the file.
3. When an unpacked file is then processed, identical block is found to be already stored and, therefore, it will be skipped.

### Pros and Cons
+ space saved
- increased time consumption due to decompression of archives (but this should run only once)

**Comment by slurdge** [comment]
It would be quite complicated to implement such a feature, because if you remove the archive and not the extracted file (from your source), then you have to "remember" that this specific file didn't exists really. Also, suppose you have an archive with 10G in it and you keep only a 1M file. You would waste 10G on backup destination.

The benefits are not worth it IMO.

**Comment by besoft** [comment]
No, you got it probably wrong. There should be no waste on backup destination.

If  you have an archive with 10G and 1M file from it elsewhere and you decide to backup the file only, then it would store just the file (1M). Similarly, for the situation when you decide to backup only the archive. BUT if  you have an archive with 10G and 1M file from it elsewhere and you decide to backup both, then it would store just the archive (10G). Suppose that this is version 0. Now you remove the archive (without chaning the file) and run backup once again. If version 0 is to keep, there is no problem. The file is already in your backup (v0) and will not be stored (just its "reference"). There is no waste on backup destination. Let say that Duplicati is configured to keep only n versions. Then the problem is what to do when version v0 should be deleted. You would need to update all versions v1 up to vn-1 that contains any reference to the archive in v0. If you do it the same way as it is done with the regular files, then yes, the archive from v0 would move to v1 and you would waste 10G. But I do not see why you could not move just the files from the archive that are referenced in version v1 up to vn-1. It won't be simple to implement, I agree but should not be killing.
